File: app/src/main/java/igrek/songbook/view/songselection/SongListView.java
package igrek.songbook.view.songselection;

import android.content.Context;
import android.graphics.Canvas;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import java.util.HashMap;
import java.util.List;

import igrek.songbook.dagger.DaggerIoc;
import igrek.songbook.logger.Logger;
import igrek.songbook.logger.LoggerFactory;
import igrek.songbook.service.songtree.SongTreeItem;

public class SongListView extends ListView implements AdapterView.OnItemClickListener, AdapterView.OnItemLongClickListener {
	
	private Logger logger = LoggerFactory.getLogger();
	private List<SongTreeItem> items;
	private SongListItemAdapter adapter;
	private HashMap<Integer, Integer> itemHeights = new HashMap<>();
	private OnSongClickListener onClickListener;
	
	{
		DaggerIoc.getFactoryComponent().inject(this);
	}
	
	public SongListView(Context context) {
		super(context);
	}
	
	public SongListView(Context context, AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
	}
	
	public SongListView(Context context, AttributeSet attrs) {
		super(context, attrs);
	}
	
	public void init(Context context, OnSongClickListener onClickListener) {
		this.onClickListener = onClickListener;
		setOnItemClickListener(this);
		setOnItemLongClickListener(this);
		setChoiceMode(ListView.CHOICE_MODE_SINGLE);
		adapter = new SongListItemAdapter(context, null, this);
		setAdapter(adapter);
	}
	
	@Override
	public boolean onTouchEvent(MotionEvent event) {
		return super.onTouchEvent(event);
	}
	
	@Override
	public void invalidate() {
		super.invalidate();
	}
	
	@Override
	protected void dispatchDraw(Canvas canvas) {
		super.dispatchDraw(canvas);
	}
	
	@Override
	public void onItemClick(AdapterView<?> parent, final View view, int position, long id) {
		SongTreeItem item = adapter.getItem(position);
		if (onClickListener != null)
			onClickListener.onSongItemClick(item);
	}
	
	@Override
	public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
		SongTreeItem item = adapter.getItem(position);
		if (onClickListener != null)
			onClickListener.onSongItemClick(item);
		return true;
	}
	
	public void setItems(List<SongTreeItem> items) {
		this.items = items;
		adapter.setDataSource(items);
		invalidate();
		calculateViewHeights();
	}
	
	private void calculateViewHeights() {
		int measureSpecW = MeasureSpec.makeMeasureSpec(this.getWidth(), MeasureSpec.EXACTLY);
		int measureSpecH = MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED);
		itemHeights = new HashMap<>();
		for (int i = 0; i < adapter.getCount(); i++) {
			View mView = adapter.getView(i, null, this);
			mView.measure(measureSpecW, measureSpecH);
			itemHeights.put(i, mView.getMeasuredHeight());
		}
	}
	
	private int getRealScrollPosition() {
		if (getChildAt(0) == null)
			return 0;
		int sumh = 0;
		for (int i = 0; i < getFirstVisiblePosition(); i++) {
			sumh += getItemHeight(i);
		}
		return sumh - getChildAt(0).getTop();
	}
	
	private int getItemHeight(int position) {
		Integer h = itemHeights.get(position);
		if (h == null)
			logger.warn("Item View = null");
		return h != null ? h : 0;
	}
	
	/**
	 * @param position position of element to scroll (-1 - last element)
	 */
	public void scrollTo(int position) {
		if (position == -1)
			position = items.size() - 1;
		if (position < 0)
			position = 0;
		setSelection(position);
		invalidate();
	}
	
	public void scrollToPosition(int y) {
		// determine nearest element and scroll relative to it
		int position = 0;
		while (y > itemHeights.get(position)) {
			y -= itemHeights.get(position);
			position++;
		}
		
		setSelection(position);
		smoothScrollBy(y, 50);
		
		invalidate();
	}
	
	public Integer getCurrentScrollPosition() {
		return getRealScrollPosition();
	}
}

File: app/src/main/java/igrek/songbook/songselection/listview/SongListView.kt
package igrek.songbook.songselection.listview

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ListView
import igrek.songbook.songselection.SongClickListener
import igrek.songbook.songselection.contextmenu.SongContextMenuBuilder
import igrek.songbook.songselection.tree.SongTreeItem

class SongListView : ListView, AdapterView.OnItemClickListener,
    AdapterView.OnItemLongClickListener {

    private var adapter: SongListItemAdapter? = null
    private var onClickListener: SongClickListener? = null

    val currentScrollPosition: ListScrollPosition
        get() {
            var yOffset = 0
            if (childCount > 0) {
                yOffset = -getChildAt(0).top
            }
            return ListScrollPosition(firstVisiblePosition, yOffset)
        }

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    )

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    fun init(
        context: Context,
        onClickListener: SongClickListener,
        songContextMenuBuilder: SongContextMenuBuilder
    ) {
        this.onClickListener = onClickListener
        onItemClickListener = this
        onItemLongClickListener = this
        choiceMode = CHOICE_MODE_SINGLE
        itemsCanFocus = true
        isFocusable = true
        descendantFocusability = ViewGroup.FOCUS_BEFORE_DESCENDANTS
        adapter = SongListItemAdapter(context, emptyList(), songContextMenuBuilder)
        setAdapter(adapter)
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val item = adapter!!.getItem(position)
        if (onClickListener != null)
            onClickListener!!.onSongItemClick(item!!)
    }

    override fun onItemLongClick(
        parent: AdapterView<*>,
        view: View,
        position: Int,
        id: Long
    ): Boolean {
        val item = adapter!!.getItem(position)
        if (onClickListener != null)
            onClickListener!!.onSongItemLongClick(item!!)
        return true
    }

    fun setItems(items: List<SongTreeItem>) {
        adapter!!.setDataSource(items)
        invalidate()
    }

    /**
     * @param position of element to scroll
     */
    private fun scrollTo(position: Int) {
        setSelection(position)
        invalidate()
    }

    fun scrollToBeginning() {
        scrollTo(0)
    }

    fun restoreScrollPosition(scrollPosition: ListScrollPosition?) {
        if (scrollPosition != null) {
            // scroll to first position
            setSelection(scrollPosition.firstVisiblePosition)
            // and move a little by y offset
            smoothScrollBy(scrollPosition.yOffsetPx, 50)
            invalidate()
        }
    }
}
